client/src/send_and_confirm_transactions_in_parallel.rs

```diff
@@ -0,0 +1,417 @@
+        transaction_map.insert(
+            signature,
+            TransactionData {
+                index: *index,
+                transaction,
+                last_valid_blockheight: blockhashdata.last_valid_blockheight,
+            },
+        );
+
+        if let Some(progress_bar) = progress_bar {
+            set_message_for_confirmed_transactions(
+                progress_bar,
+                confirmed_transactions.load(std::sync::atomic::Ordering::Relaxed),
+                total_transactions,
+                None,
+                blockhashdata.last_valid_blockheight,
+                &format!(
+                    "Sending {}/{} transactions",
+                    counter + 1,
+                    current_transaction_count,
+                ),
+            );
+        }
+    }
+    Ok(())
+}
+
+async fn confirm_transactions_till_block_height_and_resend_unexpired_transaction_over_tpu(
+    progress_bar: &Option<indicatif::ProgressBar>,
+    tpu_client: &TpuClient<QuicPool, QuicConnectionManager, QuicConfig>,
+    transaction_map: Arc<DashMap<Signature, TransactionData>>,
+    current_blockheight: Arc<AtomicU64>,
+    confirmed_transactions: Arc<AtomicU32>,
+    wait_till_blockheight: u64,
+    total_transactions: usize,
+) {
+    let transactions_to_confirm = transaction_map.len();
+
+    if let Some(progress_bar) = progress_bar {
+        set_message_for_confirmed_transactions(
+            progress_bar,
+            confirmed_transactions.load(std::sync::atomic::Ordering::Relaxed),
+            total_transactions,
+            Some(current_blockheight.load(Ordering::Relaxed)),
+            wait_till_blockheight,
+            &format!("Waiting for next block, {transactions_to_confirm} transactions pending..."),
+        );
+    }
+
+    // wait till all transactions are confirmed or we have surpassed max processing age for the last sent transaction
+    while !transaction_map.is_empty()
+        && current_blockheight.load(Ordering::Relaxed) < wait_till_blockheight
+    {
+        let blockheight = current_blockheight.load(Ordering::Relaxed);
+
+        if let Some(progress_bar) = progress_bar {
+            set_message_for_confirmed_transactions(
+                progress_bar,
+                confirmed_transactions.load(std::sync::atomic::Ordering::Relaxed),
+                total_transactions,
+                Some(blockheight),
+                wait_till_blockheight,
+                "Checking transaction status...",
+            );
+        }
+
+        let instant = Instant::now();
+        // retry sending transaction only over TPU port / any transactions sent over RPC will be automatically rebroadcast by the RPC server
+        let txs_to_resend_over_tpu = transaction_map
+            .iter()
+            .filter(|x| blockheight < x.last_valid_blockheight)
+            .map(|x| {
+                serialize(&x.transaction)
+                    .expect("Should serialize as it has been serialized before")
```

Comment:
could just cache the serialized tx in `TransactionData`

Reply:
This is because we then use the transaction to get the message and then send them again, else we have to store both the message and serialized transaction data which is not memory efficient.

Reply:
meh. it's like 2.5kb/tx max  :shrug: